lib/Service/BotService.php

```diff
@@ -0,0 +1,285 @@
+		$bots = $this->getBotsForToken($event->getRoom()->getToken());
+		if (empty($bots)) {
+			return;
+		}
+
+		$message = $messageParser->createMessage(
+			$event->getRoom(),
+			$event->getParticipant(),
+			$event->getComment(),
+			$this->l10nFactory->get('spreed', 'en', 'en')
+		);
+		$messageParser->parseMessage($message);
+		$messageData = [
+			'message' => $message->getMessage(),
+			'parameters' => $message->getMessageParameters(),
+		];
+
+		$attendee = $event->getParticipant()->getAttendee();
+
+		$this->sendAsyncRequests($bots, [
+			'type' => 'Create',
+			'actor' => [
+				'type' => 'Person',
+				'id' => $attendee->getActorType() . '/' . $attendee->getActorId(),
+				'name' => $attendee->getDisplayName(),
+			],
+			'object' => [
+				'type' => 'Note',
+				'id' => $event->getComment()->getId(),
+				'name' => 'message',
+				'content' => json_encode($messageData),
+				'mediaType' => 'text/markdown', // FIXME or text/plain when markdown is disabled
+			],
+			'target' => [
+				'type' => 'Collection',
+				'id' => $event->getRoom()->getToken(),
+				'name' => $event->getRoom()->getName(),
+			]
+		]);
+	}
+
+	public function afterSystemMessageSent(ChatEvent $event, MessageParser $messageParser): void {
+		$bots = $this->getBotsForToken($event->getRoom()->getToken());
+		if (empty($bots)) {
+			return;
+		}
+
+		$message = $messageParser->createMessage(
+			$event->getRoom(),
+			null,
+			$event->getComment(),
+			$this->l10nFactory->get('spreed', 'en', 'en')
+		);
+		$messageParser->parseMessage($message);
+		$messageData = [
+			'message' => $message->getMessage(),
+			'parameters' => $message->getMessageParameters(),
+		];
+
+		$this->sendAsyncRequests($bots, [
+			'type' => 'Activity',
+			'actor' => [
+				'type' => 'Person',
+				'id' => $message->getActorType() . '/' . $message->getActorId(),
+				'name' => $message->getActorDisplayName(),
+			],
+			'object' => [
+				'type' => 'Note',
+				'id' => $event->getComment()->getId(),
+				'name' => $message->getMessageRaw(),
+				'content' => json_encode($messageData),
+				'mediaType' => 'text/markdown',
+			],
+			'target' => [
+				'type' => 'Collection',
+				'id' => $event->getRoom()->getToken(),
+				'name' => $event->getRoom()->getName(),
+			]
+		]);
+	}
+
+	/**
+	 * @param Bot[] $bots
+	 * @param array $body
+	 */
+	protected function sendAsyncRequests(array $bots, array $body): void {
+		$jsonBody = json_encode($body, JSON_THROW_ON_ERROR);
+
+		foreach ($bots as $bot) {
+			$botServer = $bot->getBotServer();
+			$random = $this->secureRandom->generate(64);
+			$hash = hash_hmac('sha256', $random . $jsonBody, $botServer->getSecret());
+			$headers = [
+				'Content-Type' => 'application/json',
+				'X-Nextcloud-Talk-Random' => $random,
+				'X-Nextcloud-Talk-Signature' => $hash,
+				'X-Nextcloud-Talk-Backend' => rtrim($this->serverConfig->getSystemValueString('overwrite.cli.url'), '/') . '/',
+				'OCS-APIRequest' => 'true', // FIXME optional?
```

Comment:
2 others FIXME. Something to do?

Reply:
Due to the lack of time we currently don't add options for both, but basically should introduce them at some point